jsonrpc/filter_manager.go

```diff
@@ -226,197 +296,300 @@ func (f *FilterManager) Run() {
-	defer f.lock.Unlock()
-
-	// first include all the new headers in the blockstream for the block filters
-	for _, header := range evnt.NewChain {
-		f.blockStream.push(header)
-	}
-
-	processBlock := func(h *types.Header, removed bool) error {
-		// get the logs from the transaction
-		receipts, err := f.store.GetReceiptsByHash(h.Hash)
-		if err != nil {
-			return err
-		}
-
-		for indx, receipt := range receipts {
-			// check the logs with the filters
-			for _, log := range receipt.Logs {
-				for _, f := range f.filters {
-					if f.isLogFilter() {
-						if f.logFilter.Match(log) {
-							nn := &Log{
-								Address:     log.Address,
-								Topics:      log.Topics,
-								Data:        argBytes(log.Data),
-								BlockNumber: argUint64(h.Number),
-								BlockHash:   h.Hash,
-								TxHash:      receipt.TxHash,
-								TxIndex:     argUint64(indx),
-								Removed:     removed,
-							}
-							f.logs = append(f.logs, nn)
-						}
-					}
-				}
-			}
-		}
-
-		return nil
+// NewBlockFilter adds new BlockFilter
+func (f *FilterManager) NewBlockFilter(ws wsConn) string {
+	filter := &blockFilter{
+		filterBase: newFilterBase(ws),
+		block:      f.blockStream.Head(),
 	}
 
-	// process old chain
-	for _, i := range evnt.OldChain {
-		if processErr := processBlock(i, true); processErr != nil {
-			f.logger.Error(fmt.Sprintf("Unable to process block, %v", processErr))
-		}
-	}
-	// process new chain
-	for _, i := range evnt.NewChain {
-		if processErr := processBlock(i, false); processErr != nil {
-			f.logger.Error(fmt.Sprintf("Unable to process block, %v", processErr))
-		}
-	}
+	return f.addFilter(filter)
+}
 
-	// flush all the websocket values
-	for _, filter := range f.filters {
-		if filter.isWS() {
-			if flushErr := filter.flush(); flushErr != nil {
-				f.logger.Error(fmt.Sprintf("Unable to process flush, %v", flushErr))
-			}
-		}
+// NewLogFilter adds new LogFilter
+func (f *FilterManager) NewLogFilter(logQuery *LogQuery, ws wsConn) string {
+	filter := &logFilter{
+		filterBase: newFilterBase(ws),
+		query:      logQuery,
 	}
 
-	return nil
+	return f.addFilter(filter)
 }
 
+// Exists checks the filter with given ID exists
 func (f *FilterManager) Exists(id string) bool {
-	f.lock.Lock()
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
 	_, ok := f.filters[id]
-	f.lock.Unlock()
 
 	return ok
 }
 
-var errFilterDoesNotExists = fmt.Errorf("filter does not exists")
-
+// GetFilterChanges returns the updates of the filter with given ID in string
 func (f *FilterManager) GetFilterChanges(id string) (string, error) {
-	f.lock.Lock()
-	defer f.lock.Unlock()
+	f.lock.RLock()
+	defer f.lock.RUnlock()
 
-	item, ok := f.filters[id]
-	if !ok {
-		return "", errFilterDoesNotExists
-	}
+	filter, ok := f.filters[id]
 
-	if item.isWS() {
-		// we cannot get updates from a ws filter with getFilterChanges
-		return "", errFilterDoesNotExists
+	// we cannot get updates from a ws filter with getFilterChanges
+	if !ok || filter.isWS() {
```

Comment:
Can we split this conditional into 2 separate errors? I think we'll find it easier to debug in the long run

Reply:
Fixed at b0728d05324cea09366d9761c7661205e4e6dcac